README.md

# PRSbils  



## Updates  

![Update log](NEWS.md)  

## Introduction  

Polygenic risk score with bilevel continuous shrinkage for incorporating functional annotations.  



## How to use

1. Download/Clone PRSbils (this repository) to your local machine
2. Install required python packages: `pip3 install -r requirements.txt`
3. Get the input files ready
  - Download and extract LD reference panel from 1000 Genomes Project phase 3 (the same procedure as in [PRSCS](https://github.com/getian107/PRScs/blob/master/README.md)):
    - [African](https://www.dropbox.com/s/mq94h1q9uuhun1h/ldblk_1kg_afr.tar.gz?dl=0 "African")
    - [American](https://www.dropbox.com/s/uv5ydr4uv528lca/ldblk_1kg_amr.tar.gz?dl=0 "American")
    - [East Asian](https://www.dropbox.com/s/7ek4lwwf2b7f749/ldblk_1kg_eas.tar.gz?dl=0 "East Asian")
    - [European](https://www.dropbox.com/s/mt6var0z96vb6fv/ldblk_1kg_eur.tar.gz?dl=0 "European")
    - [South-eastern Asian](https://www.dropbox.com/s/hsm0qwgyixswdcv/ldblk_1kg_sas.tar.gz?dl=0 "South-eastern Asian")
  - Genotype file
  - GWAS summary statistics
  - Mapping file from SNPs to annotation groups
4. Run `PRSbils.py` to get the trained weights for SNPs with different functional annotations.

## Sample command

You can test the program using the following bash command after downloading the 1000G European LD referenfce panel file under the home directory:

```
pip3 install -r requirements.txt

python3 PRSbils.py \
        --ref_dir=./1000G/ldblk_1kg_eur \
        --bim_prefix=./data/genotypes_plink_chr1_train \
        --sst_file=./data/sumstat.txt \
        --map_file=./data/snpmap.txt \
        --n_gwas=50000 \
        --out_dir=./data/output_trained_weights_prsbils \
        --chrom=1 \
        --n_iter=1000 \
        --thres=1e-04 \
        --beta_std=False \
        --flip=True

```

The output file is  `./data/output_trained_weights_prsbils_beta_chr1.txt`, which is a tab-delimited text file with the following columns:
```
[Chromosome] [SNP ID] [SNP Position] [Allele 1] [Allele 2] [Weight in the summary statistics file] [Weight from PRSbils] [Annotation Group] [Subgroup (not yet implemented)] [Allele Frequency]
```

Please check the next section or run `python3 PRSbils.py --help` for details about the parameters.

## Details about the parameters

```
usage: PRSbils.py [-h] --ref_dir REF_DIR --bim_prefix BIM_PREFIX --sst_file
                  SST_FILE --map_file MAP_FILE [--map_gcol MAP_GCOL] --n_gwas
                  N_GWAS --out_dir OUT_DIR [--a A] [--b B] [--c C] [--d D]
                  [--e E] [--f F] [--n_iter N_ITER]
                  [--chrom CHROM [CHROM ...]] [--beta_std [BETA_STD]]
                  [--thres THRES] [--ignore_blk [IGNORE_BLK]]
                  [--verbose [VERBOSE]] [--fixtau [FIXTAU]] [--flip [FLIP]]

PRSbils: Polygenic Risk Score with Bilevel Continuous Shrinkage

optional arguments:
  -h, --help            show this help message and exit
  --ref_dir REF_DIR     Full path (including folder name) to the directory
                        (ldblk_1kg_eur, ldblk_1kg_eas or ldblk_1kg_afr) that
                        contains information on the LD reference panel
                        (snpinfo_1kg_hm3 and ldblk_1kg_chr*.hdf5)
  --bim_prefix BIM_PREFIX
                        Full path and the prefix of the bim file
  --sst_file SST_FILE   Full path and the file name of the GWAS summary
                        statistics Summary statistics file must be tab
                        delimited with the following format (including the
                        header line): SNP A1 A2 BETA P Or: SNP A1 A2 OR P
  --map_file MAP_FILE   Full path and the file name of the mapping file from
                        SNPs to sets. Map file must have the MSigDB file
                        format : [CHROM] [SNPID] [SETNAME] [GROUP (optional)]
  --map_gcol MAP_GCOL   (Not yet implemented) Column id for group info in map_file. Default is None
                        (no group information)
  --n_gwas N_GWAS       Sample size of the GWAS
  --out_dir OUT_DIR     Output directory and output filename prefix of the
                        posterior effect size estimates
  --a A                 Parameter a for sigma. Default is 0.001
  --b B                 Parameter b in the gamma prior for sigma. Default is
                        0.001
  --c C                 Parameter c for group implementation. Default is 0.001
  --d D                 Parameter d for group implementation. Default is 0.001
  --e E                 Parameter e for lambda. Default is 1
  --f F                 Parameter f for lambda. Default is 0.5
  --n_iter N_ITER       Number of maximum iterations. Default is 1000
  --chrom CHROM [CHROM ...]
                        The chromosome on which the model is fitted. Default
                        is iterating through 22 autosomes
  --beta_std [BETA_STD]
                        Boolean. If True, return standardized posterior SNP
                        effect sizes. Default is False
  --thres THRES         Threshold value for convergence. Default is 0.001
  --ignore_blk [IGNORE_BLK]
                        Boolean. If false, elements categorized as same set
                        but in different ld will be treated as different sets.
                        Default is False
  --verbose [VERBOSE]   Boolean. Default is False
  --fixtau [FIXTAU]     Boolean. If false, will estimate tau in algorithm,
                        otherwise will fix at 1. Default is False
  --flip [FLIP]         Boolean. If false, signs of beta in summary statistics
                        will not be flipped when there's a switch in A1 and A2
                        alleles. Default is False
```


NEWS.md
## News
2022-12-12: Updated readme file and examples.
2022-08-08: Pre-release initial commit.

test/example.sh
#!/bin/sh

pip3 install -r requirements.txt

python3 PRSbils.py --ref_dir=./1000G/ldblk_1kg_eur --bim_prefix=./data/genotypes_plink_chr1_train --sst_file=./data/sumstat.txt --map_file=./data/snpmap.txt --n_gwas=50000 --out_dir=./data/output_trained_weights_prsbils --chrom=1 --n_iter=1000 --thres=1e-04 --beta_std=False --flip=True


test/__init__.py
from unittest import TestCase

import prsbils

class TestPRSbils(TestCase):
    def test_is_string(self):
        s = funniest.joke()
        self.assertTrue(isinstance(s, basestring))

utils.py
#!/usr/bin/env python3

"""
Utility functions for general data transformation.

"""
import numpy as np
import scipy.special as sc
from scipy.linalg import block_diag
import pandas as pd
import statistics
import math
from sklearn.linear_model import LogisticRegression
from itertools import combinations

def mergeDict(dict1, dict2, mode="matrix", groups=None):
    ''' 
    Merge dictionaries in order dict1, dict2
    @params    dict1
    @params dict2
    @params    mode     "matrix": form block diagonal matrix; "list": append list in dict2 to that in dict1 with same key; "numeric": sum numeric elements with same key
    @example
    test1 = {'a':[np.array([[1,2],[3,4]])]}
    test2 = {'a':[np.array([[5,6],[7,8]])]}
    test3 = utils.mergeDict(test1, test2, mode="matrix")

    test1 = {'a':[1,2], 'b':[3,4]}
    test2 = {'a':[5,6],'c':[7,8]}
    test3 = utils.mergeDict(test1, test2, mode="list")

    test1 = {'a':{'a':[1,2], 'b':[3,4]}, 'b':{'a':[1,2], 'b':[3,4]}}
    test2 = {'a':{'a':[5,6], 'b':[7,8]},'c':{'a':[9,10], 'b':[11,12]}}
    test3 = utils.mergeDict(test1, test2, mode="list", groups = set(['a','b']))
    '''
    # dict3 = {**dict1, **dict2}
    dict3 = dict2.copy()
    dict3.update(dict1)

    if groups == None:
        if mode=="matrix":
            for key, value in dict3.items():
                if key in dict1 and key in dict2:
                    dict3[key]=block_diag(dict3[key],dict2[key])
        elif mode =="list":
            for key, value in dict3.items():
                if key in dict1 and key in dict2:
                    dict3[key]= np.append(dict3[key], dict2[key])
                    # dict3[key].append(dict2[key])
        elif mode =="numeric":
            for key, value in dict3.items():
                if key in dict1 and key in dict2:
                    dict3[key] = dict3[key] + dict2[key]
    else: 
        # if groups not none
        for gg in groups:
            if mode=="matrix":
                for key, value in dict3.items():
                    if key in dict1 and key in dict2:
                        dict3[key][gg]=block_diag(dict3[key][gg],dict2[key][gg])
            elif mode =="list":
                for key, value in dict3.items():
                    if key in dict1 and key in dict2:
                        dict3[key][gg]= np.append(dict3[key][gg], dict2[key][gg])
                        # dict3[key].append(dict2[key])
            elif mode =="numeric":
                for key, value in dict3.items():
                    if key in dict1 and key in dict2:
                        dict3[key][gg] = dict3[key][gg] + dict2[key][gg]

    return dict3

def getOR(x,y, by_decile=10):
    x_array = np.squeeze(np.asarray(x))
    y_array = np.squeeze(np.asarray(y))
    if by_decile:
        deciles = pd.qcut(x_array,by_decile)
        deciles_dum = pd.get_dummies(data=deciles, drop_first=True)
        # dat = np.column_stack((x,deciles_dum))
        onevec = np.ones(len(y_array))
        dumvec = deciles_dum.values[:,-1]
        # dat = deciles_dum
        dat = np.column_stack((onevec, dumvec))
        glm_model = LogisticRegression(random_state=0, solver='lbfgs',fit_intercept=False).fit(dat, y_array)
        coef = glm_model.coef_[0]
        # out = [math.exp(c) for c in coef]
        out = math.exp(coef[-1])
        return(out)

    else:
        print('Need to specify number of deciles')
        return(0)

def getR2(x,y, by_decile=0):
    x_array = np.squeeze(np.asarray(x))
    # x_array = (x_array - mean(x_array)) / std(x_array)
    y_array = np.squeeze(np.asarray(y))
    if by_decile:
        deciles = pd.qcut(x_array,by_decile)
        cat = deciles.categories
        r2 = []
        for c in cat:
            idx = np.where(deciles==c)
            corr_matrix = np.corrcoef(x_array[idx], y_array[idx])
            corr_xy = corr_matrix[0,1]
            r2.append(corr_xy**2)
        return(r2)
    else:
        correlation_matrix = np.corrcoef(x_array, y_array)
        correlation_xy = correlation_matrix[0,1]
        r2 = correlation_xy**2
        return(r2)

def getIOU(x,y):
    """
    Intersection over Union for two sets
    """
    set1 = set(x)
    set2 = set(y)
    m1 = len(set1)
    m2 = len(set2)
    m12 = len(set1 & set2)
    iou = m12/(m1+m2-m12)
    return(iou)

def pairwise(l):
    """Get an iterable pair from list"""
    out = list(combinations(l, 2))
    return out

def _lnexp1_taylor(x, n_term):
    """Series expansion for E1"""
    r = 0; s = -1
    for i in range(n_term, 0, -1):
        r = (r + s) * i / x
        s = -s
    return r*s

def lnexp1(x, n_term=18):
    """Calculation of log of Exponential integral E1 using series expansion"""
    x = np.array(x)
    use_approx = x >= 50

    ax = np.where(use_approx, x, 100)
    sx = np.where(use_approx, 1, x)
    approx = 0.0-x -np.log(x) + np.log1p(_lnexp1_taylor(ax, n_term))
    temp = np.log(sc.exp1(sx))
    return np.where(use_approx, approx, temp) * 1

def safeexp(x, floor=-710,ceil=710):
    if x <= floor:
        out = 0
    elif x >=ceil:
        out = np.exp(ceil-1)
    else:
        out = np.exp(x)
    return out

def multiReplace(indata, dictionary):
    for key in dictionary.keys():
        indata = [aa.replace(key, dictionary[key]) for aa in indata]
    output = [aa.upper() for aa in indata]
    return output

def multiReplace1(indata, dictionary):
    for key in dictionary.keys():
        indata = indata.replace(key, dictionary[key])
    output = indata.upper()
    return output

gigrnd.py
#!/usr/bin/env python3

""" Title: PRScs source code, parse_ref
    Author: Tian Ge
    Date: 2021
    Code version: 1.0.0
    Availability: https://github.com/getian107/PRScs
    Description: Random variate generator for the generalized inverse Gaussian distribution.
    Reference: L Devroye. Random variate generation for the generalized inverse Gaussian distribution.
           Statistics and Computing, 24(2):239-246, 2014.
"""
import math
from scipy import random


def psi(x, alpha, lam):
    f = -alpha*(math.cosh(x)-1.0)-lam*(math.exp(x)-x-1.0)
    return f


def dpsi(x, alpha, lam):
    f = -alpha*math.sinh(x)-lam*(math.exp(x)-1.0)
    return f


def g(x, sd, td, f1, f2):
    if (x >= -sd) and (x <= td):
        f = 1.0
    elif x > td:
        f = f1
    elif x < -sd:
        f = f2

    return f


def gigrnd(p, a, b):
    # setup -- sample from the two-parameter version gig(lam,omega)
    p = float(p); a = float(a); b = float(b)
    lam = p
    omega = math.sqrt(a*b)

    if lam < 0:
        lam = -lam
        swap = True
    else:
        swap = False

    alpha = math.sqrt(math.pow(omega,2)+math.pow(lam,2))-lam

    # find t
    x = -psi(1.0, alpha, lam)
    if (x >= 0.5) and (x <= 2.0):
        t = 1.0
    elif x > 2.0:
        if (alpha == 0) and (lam == 0):
            t = 1.0
        else:
            t = math.sqrt(2.0/(alpha+lam))
    elif x < 0.5:
        if (alpha == 0) and (lam == 0):
            t = 1.0
        else:
            t = math.log(4.0/(alpha+2.0*lam))

    # find s
    x = -psi(-1.0, alpha, lam)
    if (x >= 0.5) and (x <= 2.0):
        s = 1.0
    elif x > 2.0:
        if (alpha == 0) and (lam == 0):
            s = 1.0
        else:
            s = math.sqrt(4.0/(alpha*math.cosh(1)+lam))
    elif x < 0.5:
        if (alpha == 0) and (lam == 0):
            s = 1.0
        elif alpha == 0:
            s = 1.0/lam
        elif lam == 0:
            s = math.log(1.0+1.0/alpha+math.sqrt(1.0/math.pow(alpha,2)+2.0/alpha))
        else:
            s = min(1.0/lam, math.log(1.0+1.0/alpha+math.sqrt(1.0/math.pow(alpha,2)+2.0/alpha)))

    # find auxiliary parameters
    eta = -psi(t, alpha, lam)
    zeta = -dpsi(t, alpha, lam)
    theta = -psi(-s, alpha, lam)
    xi = dpsi(-s, alpha, lam)

    p = 1.0/xi
    r = 1.0/zeta

    td = t-r*eta
    sd = s-p*theta
    q = td+sd

    # random variate generation
    while True:
        U = random.random()
        V = random.random()
        W = random.random()
        if U < q/(p+q+r):
            rnd = -sd+q*V
        elif U < (q+r)/(p+q+r):
            rnd = td-r*math.log(V)
        else:
            rnd = -sd+p*math.log(V)

        f1 = math.exp(-eta-zeta*(rnd-t))
        f2 = math.exp(-theta+xi*(rnd+s))
        if W*g(rnd, sd, td, f1, f2) <= math.exp(psi(rnd, alpha, lam)):
            break

    # transform back to the three-parameter version gig(p,a,b)
    rnd = math.exp(rnd)*(lam/omega+math.sqrt(1.0+math.pow(lam,2)/math.pow(omega,2)))
    if swap:
        rnd = 1.0/rnd

    rnd = rnd/math.sqrt(a/b)
    return rnd


PRSbils.py
#!/usr/bin/env python3

"""
usage: PRSbils.py [-h] --ref_dir REF_DIR --bim_prefix BIM_PREFIX --sst_file
                  SST_FILE --map_file MAP_FILE [--map_gcol MAP_GCOL] --n_gwas
                  N_GWAS --out_dir OUT_DIR [--a A] [--b B] [--c C] [--d D]
                  [--e E] [--f F] [--n_iter N_ITER]
                  [--chrom CHROM [CHROM ...]] [--beta_std [BETA_STD]]
                  [--thres THRES] [--ignore_blk [IGNORE_BLK]]
                  [--verbose [VERBOSE]] [--fixtau [FIXTAU]] [--flip [FLIP]]

PRSbils: Polygenic Risk Score with Bilevel Continuous Shrinkage

optional arguments:
  -h, --help            show this help message and exit
  --ref_dir REF_DIR     Full path (including folder name) to the directory
                        (ldblk_1kg_eur, ldblk_1kg_eas or ldblk_1kg_afr) that
                        contains information on the LD reference panel
                        (snpinfo_1kg_hm3 and ldblk_1kg_chr*.hdf5)
  --bim_prefix BIM_PREFIX
                        Full path and the prefix of the bim file
  --sst_file SST_FILE   Full path and the file name of the GWAS summary
                        statistics Summary statistics file must be tab
                        delimited with the following format (including the
                        header line): SNP A1 A2 BETA P Or: SNP A1 A2 OR P
  --map_file MAP_FILE   Full path and the file name of the mapping file from
                        SNPs to sets. Map file must have the MSigDB file
                        format : [CHROM] [SNPID] [SETNAME] [GROUP (optional)]
  --map_gcol MAP_GCOL   (Not yet implemented) Column id for group info in map_file. Default is None
                        (no group information)
  --n_gwas N_GWAS       Sample size of the GWAS
  --out_dir OUT_DIR     Output directory and output filename prefix of the
                        posterior effect size estimates
  --a A                 Parameter a for sigma. Default is 0.001
  --b B                 Parameter b in the gamma prior for sigma. Default is
                        0.001
  --c C                 Parameter c for group implementation. Default is 0.001
  --d D                 Parameter d for group implementation. Default is 0.001
  --e E                 Parameter e for lambda. Default is 1
  --f F                 Parameter f for lambda. Default is 0.5
  --n_iter N_ITER       Number of maximum iterations. Default is 1000
  --chrom CHROM [CHROM ...]
                        The chromosome on which the model is fitted. Default
                        is iterating through 22 autosomes
  --beta_std [BETA_STD]
                        Boolean. If True, return standardized posterior SNP
                        effect sizes. Default is False
  --thres THRES         Threshold value for convergence. Default is 0.001
  --ignore_blk [IGNORE_BLK]
                        Boolean. If false, elements categorized as same set
                        but in different ld will be treated as different sets.
                        Default is False
  --verbose [VERBOSE]   Boolean. Default is False
  --fixtau [FIXTAU]     Boolean. If false, will estimate tau in algorithm,
                        otherwise will fix at 1. Default is False
  --flip [FLIP]         Boolean. If false, signs of beta in summary statistics
                        will not be flipped when there's a switch in A1 and A2
                        alleles. Default is False
"""

import sys
import argparse
import time
import parseg
import vbclass_mcmc
import gigrnd

def str2bool(v):
    if isinstance(v, bool):
       return v
    if v.lower() in ('yes', 'true', 't', 'y', '1'):
        return True
    elif v.lower() in ('no', 'false', 'f', 'n', '0'):
        return False
    else:
        print('Invalid Boolean argument\n')
        sys.exit(2)

def parse_param():
    parser = argparse.ArgumentParser(description='PRSbils: Polygenic Risk Score with Bilevel Continuous Shrinkage')
    parser.add_argument("--ref_dir", help="Full path (including folder name) to the directory (ldblk_1kg_eur, ldblk_1kg_eas or ldblk_1kg_afr) that contains information on the LD reference panel (snpinfo_1kg_hm3 and ldblk_1kg_chr*.hdf5)", required=True)
    parser.add_argument("--bim_prefix", help="Full path and the prefix of the bim file", required=True)
    parser.add_argument("--sst_file",
        help='''
        Full path and the file name of the GWAS summary statistics
        Summary statistics file must be tab delimited with the following format (including the header line):

        SNP          A1   A2   BETA      P

        Or:
        SNP          A1   A2   OR        P
        ''', required=True)
    parser.add_argument("--map_file", help='''
        Full path and the file name of the mapping file from SNPs to sets.
        Map file must have the MSigDB file format :

        [CHROM]    [SNPID]    [SETNAME]    [GROUP (optional)]
        ''', required=True)
    parser.add_argument("--map_gcol", type=int, default=None, help="Column id for group info in map_file. Default is None (no group information)", required=False)
    parser.add_argument("--n_gwas", type=int, help="Sample size of the GWAS", required=True)
    parser.add_argument("--out_dir", help="Output directory and output filename prefix of the posterior effect size estimates", required=True)
    parser.add_argument("--a", type=float, default=0.001, help="Parameter a for sigma. Default is 0.001", required=False)
    parser.add_argument("--b", type=float, default=0.001, help="Parameter b in the gamma prior for sigma. Default is 0.001", required=False)
    parser.add_argument("--c", type=float, default=0.001, help="Parameter c for group implementation. Default is 0.001", required=False)
    parser.add_argument("--d", type=float, default=0.001, help="Parameter d for group implementation. Default is 0.001", required=False)
    parser.add_argument("--e", type=float, default=1.0, help="Parameter e for lambda. Default is 1", required=False)
    parser.add_argument("--f", type=float, default=0.5, help="Parameter f for lambda. Default is 0.5", required=False)
    parser.add_argument("--n_iter", type=int, default=1000, help="Number of maximum iterations. Default is 1000", required=False)
    parser.add_argument("--chrom", nargs='+', type=int, default=range(1,23), help="The chromosome on which the model is fitted. Default is iterating through 22 autosomes", required=False)
    parser.add_argument("--beta_std", type=str2bool, nargs='?', const=True, default=False, help="Boolean. If True, return standardized posterior SNP effect sizes. Default is False")
    parser.add_argument("--thres", type=float, default=0.001, help="Threshold value for convergence. Default is 0.001", required=False)
    parser.add_argument("--ignore_blk", type=str2bool, nargs='?', const=True, default=False, help="Boolean. If false, elements categorized as same set but in different ld will be treated as different sets. Default is False")
    parser.add_argument("--verbose", type=str2bool, nargs='?', const=True, default=False, help="Boolean. Default is False")
    parser.add_argument("--fixtau", type=str2bool, nargs='?', const=True, default=False, help="Boolean. If false, will estimate tau in algorithm, otherwise will fix at 1. Default is False")
    parser.add_argument("--flip", type=str2bool, nargs='?', const=True, default=False, help="Boolean. If false, signs of beta in summary statistics will not be flipped when there's a switch in A1 and A2 alleles. Default is False")

    param_dict = parser.parse_args()
    return param_dict


def main():
    print('='*30 + " PRSbils " +'='*30 + '\n')
    param_dict = parse_param()
    print(param_dict)
    print('-'*69 + '\n')

    for chrom in param_dict.chrom:
        print('##### process chromosome %d #####' % int(chrom))
        
        tic = time.time()
        ref_dict = parseg.parse_ref(param_dict.ref_dir + '/snpinfo_1kg_hm3', int(chrom))
        vld_dict = parseg.parse_bim(param_dict.bim_prefix, int(chrom))
        sst_dict, comm_snp = parseg.parse_sumstats(ref_dict, vld_dict, param_dict.sst_file, param_dict.n_gwas, param_dict.flip)
        set_dict, noanno_snp = parseg.parse_map(param_dict.map_file, param_dict.map_gcol, int(chrom), comm_snp)
        ld_blk_set, snp_blk_set, blk_size_set, blk_size = parseg.parse_ldblk_set(param_dict.ref_dir, sst_dict, set_dict, int(chrom), param_dict.ignore_blk, comm_snp, noanno_snp) # creating new set for snps w/o annotation
        toc = time.time()
        print('Data-parsing completed in %0.4f seconds' % (toc - tic))
        print('-'*69 + '\n')

        groups = set(set_dict['GROUP'])

        vbobj = vbclass_mcmc.VBCS(sst_dict, ld_blk_set, snp_blk_set, blk_size_set, param_dict.n_gwas, groups, param_dict.a, param_dict.b, param_dict.c, param_dict.d, param_dict.e, param_dict.f, chrom)

        tic = time.time()
        his = vbobj.fit_mcmc(param_dict.n_iter, param_dict.thres, param_dict.verbose, 500, 1.0, param_dict.fixtau)
        toc = time.time()
        print('Parameter estimation completed in %0.4f seconds' % (toc-tic))
        print('-'*69 + '\n')

        vbobj.write_beta_mcmc(param_dict.out_dir,param_dict.beta_std)
        print('Output weight file saved to\n%s' % param_dict.out_dir)
        print('='*69 + '\n')


if __name__ == '__main__':
    main()
